pkg/cloudsql/adapter.go

```diff
@@ -0,0 +1,263 @@
+	"google.golang.org/api/sqladmin/v1"
+)
+
+type CloudSQLAdapter struct {
+	agent.CommonAgent
+	PGDriver              *pgPool.Pool
+	State                 *State
+	CloudSQLConfig        Config
+	CloudMonitoringClient *CloudMonitoringClient
+	CloudSQLAdminClient   *SqlAdminClient
+	GuardrailSettings     *guardrails.Config
+}
+
+func CreateCloudSQLAdapter() (*CloudSQLAdapter, error) {
+	ctx := context.Background()
+	commonAgent := agent.CreateCommonAgent()
+
+	pgConfig, err := pg.ConfigFromViper(nil)
+	if err != nil {
+		return nil, err
+	}
+
+	pgPool, err := pgPool.New(ctx, pgConfig.ConnectionURL)
+	if err != nil {
+		return nil, err
+	}
+
+	config, err := ConfigFromViper()
+	if err != nil {
+		return nil, err
+	}
+
+	client, err := monitoring.NewMetricClient(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	sqladminService, err := sqladmin.NewService(ctx)
+	if err != nil {
+		log.Fatalf("Failed to connect to SQL Admin service: %v", err)
+	}
+
+	guardrailSettings, err := guardrails.ConfigFromViper(nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to validate settings for guardrails %w", err)
+	}
+
+	c := &CloudSQLAdapter{
+		CommonAgent:    *commonAgent,
+		State:          &State{LastGuardrailCheck: time.Now()},
+		PGDriver:       pgPool,
+		CloudSQLConfig: config,
+		CloudMonitoringClient: &CloudMonitoringClient{
+			client: client,
+			ctx:    ctx,
+		},
+		CloudSQLAdminClient: &SqlAdminClient{
+			client: sqladminService,
+		},
+		GuardrailSettings: &guardrailSettings,
+	}
+
+	c.InitCollectors(c.Collectors())
+
+	return c, nil
+}
+
+func (adapter *CloudSQLAdapter) ApplyConfig(proposedConfig *agent.ProposedConfigResponse) error {
+	// TODO: need to make sure that the last update has been applied - I think this is fairly "edge casey"
+	// but the update is non-zero in time so we should gracefully handle
+	// also the knobs that need restarts!
+
+	adapter.Logger().Infof("Applying config")
+	adapter.Logger().Infof("Proposed Config: %v", proposedConfig)
+	adapter.Logger().Infof("Knobs to Override: %v", proposedConfig.KnobsOverrides)
+
+	flags := []*sqladmin.DatabaseFlags{}
+
+	for _, knob := range proposedConfig.KnobsOverrides {
+		knobConfig, err := parameters.FindRecommendedKnob(proposedConfig.Config, knob)
+		if err != nil {
+			return fmt.Errorf("failed to find recommended knob: %v", err)
+		}
+		fmtValue, err := knobConfig.GetSettingValue()
+		if err != nil {
+			return fmt.Errorf("failed to get setting value: %v", err)
+		}
+
+		param := &sqladmin.DatabaseFlags{
+			Name:  knobConfig.Name,
+			Value: fmtValue,
+		}
+
+		flags = append(flags, param)
+	}
+
+	adapter.CloudSQLAdminClient.applyFlags(adapter.CloudSQLConfig.ProjectID, adapter.CloudSQLConfig.DatabaseName, flags)
+
+	return nil
+}
+
+func (adapter *CloudSQLAdapter) GetActiveConfig() (agent.ConfigArraySchema, error) {
+	adapter.Logger().Debugf("Getting Active Config")
+
+	return pg.GetActiveConfig(adapter.PGDriver, context.Background(), adapter.Logger())
+}
```

Comment:
Nice, nothing specific to pull from their API for this I guess?

Reply:
So you can do both: https://cloud.google.com/sql/docs/postgres/flags#view_current_values_of_database_flags but I went for what we already had